InstanceManager.js

```diff
@@ -409,94 +425,171 @@ class InstanceManager {
-      );
+    while (true) {
+      if (instance.status !== 'RUNNING') {
+        throw new Error(`
+          Instance ${instance.name} is down!
+          Real status ${instance.status} with exitcode ${instance.exitcode}.
+          See logfile here ${instance.logFile}`);
+      }
+     
+      let _test_ = Date.now() - started;
+      
+      if (_test_ > WAIT_TIMEOUT * 1000) {
+        throw new Error(
+          `Instance ${instance.name} is still not ready after ${WAIT_TIMEOUT} secs`
+        );
+      }
+
+      let ok = false;
+      try {
+        await rp.get({
+          uri: endpointToUrl(instance.endpoint) + '/_api/version'
+        });
+        return instance;
+      } catch (e) {}
+      // Wait 100 ms and try again
+      await sleep(100);
     }
+  }
 
-    let ok = false;
-    try {
-      await rp.get({
-        uri: endpointToUrl(instance.endpoint) + '/_api/version'
+  async waitForSyncReplication() {
+    const baseUrl = endpointToUrl(this.getAgencyEndpoint());
+    // We wait at most 50s for everything to get in sync
+    for (let i = 0; i < 100; ++i) {
+      let [info] = await this.rpAgency({
+        method: 'POST',
+        uri: baseUrl + '/_api/agency/read',
+        json: true,
+        body: [
+          [
+            '/arango/Current/Collections/_system'
+          ]
+        ]
       });
-      ok = true;
-    } catch (e) {}
-    if (ok) {
-      return instance;
-    } else {
-      await sleep(100);
-      return this.waitForInstance(instance, started);
+      if (!info.hasOwnProperty('arango')
+          || !info.arango.hasOwnProperty('Current')
+          || !info.arango.Current.hasOwnProperty('Collections')
+          || !info.arango.Current.Collections.hasOwnProperty('_system')) {
+        debugLog('Got unexpected result from AgencyPlan:', info);
+        await sleep(500);
+        continue;
+      }
+      const current = info.arango.Current.Collections._system;
+
+      let foundNotInSync = false;
+      if (Object.keys(current).length < 13) {
+        // The Cluster needs to create 13 system collections. Wait and try again
+        await sleep(500);
+        continue;
+      }
+      for (const [key, collection] of Object.entries(current)) {
+        if (foundNotInSync) {
+          break;
+        }
+        for (const [sname, shard] of Object.entries(collection)) {
+          if (foundNotInSync) {
+            break;
+          }
+          if (!Array.isArray(shard.servers) || shard.servers.length < 2) {
+            // We do not have one in sync follower for each shard
+            // Sleep and try again
+            foundNotInSync = true;
+            continue;
+          }
+        }
+      }
+      if (!foundNotInSync) {
+        return true;
+      }
+      await sleep(500);
     }
   }
 
-  waitForAllInstances() {
-    return Promise.all(
-      this.instances.map(instance => {
-        return this.waitForInstance(instance);
-      })
-    );
+  async waitForAllInstances() {
```

Comment:
This is nearly identical to the method below. Can't we implement one in terms of the other? e.g.
```js
const results = await waitForInstances(this.instances);
await this.waitForSyncReplication();
return results;
```